wolfcrypt/src/hpke.c

```diff
@@ -0,0 +1,1139 @@
+        labeled_info_p += suite_id_len;
+
+        /* label */
+        XMEMCPY(labeled_info_p, label, label_len);
+        labeled_info_p += label_len;
+
+        /* info */
+        XMEMCPY(labeled_info_p, info, infoSz);
+        labeled_info_p += infoSz;
+
+        /* call expand */
+        ret = wc_HKDF_Expand(hpke->kdf_digest,
+            prk, prk_len,
+            labeled_info, (word32)(size_t)(labeled_info_p - labeled_info),
+            out, L);
+    }
+
+#ifdef WOLFSSL_SMALL_STACK
+    XFREE(labeled_info, hpke->heap, DYNAMIC_TYPE_TMP_BUFFER);
+#endif
+
+    return ret;
+}
+
+static int wc_HpkeContextComputeNonce(Hpke* hpke, HpkeBaseContext* context,
+    byte* out)
+{
+    int i;
+    int ret;
+    byte seq_bytes[HPKE_Nn_MAX];
+
+    /* convert the sequence into a byte string with the same length as the
+     * nonce */
+    ret = I2OSP(context->seq, hpke->Nn, seq_bytes);
+    if (ret == 0) {
+        for (i = 0; i < (int)hpke->Nn; i++) {
+            out[i] = (context->base_nonce[i] ^ seq_bytes[i]);
+        }
+    }
+
+    return ret;
+}
+
+static int wc_HpkeExtractAndExpand( Hpke* hpke, byte* dh, word32 dh_len,
+    byte* kemContext, word32 kem_context_length, byte* sharedSecret)
+{
+    int ret;
+    /* max length is the largest hmac digest possible */
+#ifndef WOLFSSL_SMALL_STACK
+    byte eae_prk[WC_MAX_DIGEST_SIZE];
+#else
+    byte* eae_prk;
+#endif
+
+    if (hpke == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+#ifdef WOLFSSL_SMALL_STACK
+    eae_prk = (byte*)XMALLOC(WC_MAX_DIGEST_SIZE, hpke->heap,
+        DYNAMIC_TYPE_DIGEST);
+    if (eae_prk == NULL) {
+        return MEMORY_E;
+    }
+#endif
+
+    /* extract */
+    ret = wc_HpkeLabeledExtract(hpke, hpke->kem_suite_id,
+        sizeof( hpke->kem_suite_id ), NULL, 0, (byte*)EAE_PRK_LABEL_STR,
+        EAE_PRK_LABEL_STR_LEN, dh, dh_len, eae_prk);
+
+    /* expand */
+    if ( ret == 0 )
+        ret = wc_HpkeLabeledExpand(hpke, hpke->kem_suite_id,
+            sizeof( hpke->kem_suite_id ), eae_prk, hpke->Nh,
+            (byte*)SHARED_SECRET_LABEL_STR, SHARED_SECRET_LABEL_STR_LEN,
+            kemContext, kem_context_length, hpke->Nsecret, sharedSecret);
+
+#ifdef WOLFSSL_SMALL_STACK
+    XFREE(eae_prk, hpke->heap, DYNAMIC_TYPE_DIGEST);
+#endif
+
+    return ret;
+}
+
+static int wc_HpkeKeyScheduleBase(Hpke* hpke, HpkeBaseContext* context,
+    byte* sharedSecret, byte* info, word32 infoSz)
+{
+    int ret;
+#ifndef WOLFSSL_SMALL_STACK
+    /* 1 for mode and WC_MAX_DIGEST_SIZE times 2 for psk_id_hash and */
+    /* info_hash */
+    byte key_schedule_context[1 + 2 * WC_MAX_DIGEST_SIZE];
+    /* maximum size of secret is largest hash of extract */
+    byte secret[WC_MAX_DIGEST_SIZE];
+#else
+    byte* key_schedule_context = NULL;
+    byte* secret = NULL;
+#endif
+
+    if (hpke == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+#ifdef WOLFSSL_SMALL_STACK
+    key_schedule_context = (byte*)XMALLOC((1 + 2 * WC_MAX_DIGEST_SIZE),
```

Comment:
I see +1 here and at other places too, is that for having a null terminator? Please comment on why adding 1 to the size

Reply:
its written above the variable definition `    /* 1 for mode and WC_MAX_DIGEST_SIZE times 2 for psk_id_hash and */
`